### PR TITLE
Add gperftools profiler

### LINKED ISSUE #1053: integration of a simple profiler
### Motivation / Current Behaviour

#942 added documentation of scorep profiling. scorep is very complicated and overkill for most issues, there are simple profiling tools available.

### Enhancement description

integrate at least one simple profiling tool into the build system that is reliable and easy to use.

### Additional context

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project

### PR DESCRIPTION
**Pull request by dabele**
# Changes and Information

- Integrated gperftools for profiling into cmake
- added documentation about gperftools usage and general profling

@DavidKerkmann maybe you could give gperftools on MacOS a try? I haven't been able to find reliable information about whether it works or how to get it. apparently homebrew should work?

Closes #1053 

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful (with and without OpenMP)
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [x] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [ ] Corresponding issue(s) is/are linked and addressed
- [ ] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [ ] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [ ] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [ ] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

### CODE REVIEW

File: cpp/benchmarks/profiling.md [hunk 1]
```diff
@@ -1,4 +1,39 @@
...
+- sampling: checking the CPU stack register in regular intervals to see which code is currently being executed. Advantage: usually low measurement runtime overhead. Disadvantage: stochastic results, imprecise especially for short functions, longer runtimes improve averages.
+- emulating: running the program in a virtual environment. Advantage: exact and able to measure even small functions. Disadvantage: program can run significantly longer, no real world timing results but instruction cycle counts are generally directly translatable.
+
+These tools have been tried for memilio and are described below:
+- gperftools: sampling profiler that gives reliable results with little setup
+- Score-P: large suite that supports both sampling and instrumenting. Special support for parallel code (OpenMP and MPI), including tracing to find load imbalances in the parallelization. 
+- Valgrind: suite that includes the Callgrind emulating profiler; description NYI; see [here](https://valgrind.org/info/tools.html)
```
Comment (reneSchm) [inline cpp/benchmarks/profiling.md:11]: I don't like the abbreviation NYI here (I at least don't see it used often), how about 

```suggestion
- Valgrind: suite that includes the Callgrind emulating profiler; a description can be found at the [official site](https://valgrind.org/info/tools.html)
```

or something similar. Also, using callgrind is pretty easy, if you like we could add the steps:
1. Compile with mode "RelWithDebInfo".
2. Run `valgrind --tool=callgrind \<path/to/binary\>`
3. Open "callgrind.out.\<PID\>" in KCacheGrind to visualize the call graph.

**Review by reneSchm — CHANGES_REQUESTED** [review]
I could not get this to work on the cluster, but it looks good otherwise.
In the cpp/README.md the entry for the benchmark dir is missing, maybe we could add something like
```
- benchmarks: applications to measure the performance of the framework. More information on profiling [can be found here](benchmarks/profiling.md).
```
so that the documentation is easier to find. I had to look at the "Files changed" tab to locate it.
Also, please add a description for "MEMILIO_ENABLE_PROFILING" in the "Configuring using CMake" section.

File: cpp/benchmarks/profiling.md [hunk 2]
```diff
@@ -1,4 +1,39 @@
...
+4. Test the functionality.
+5. Confirm the optimization with a benchmark and profile.
+6. Repeat if necessary.
+
+## gperftools
+
+gperftools (formerly google performance tools) is a suite of performance tools. Here the focus is on the [profiler](https://gperftools.github.io/gperftools/cpuprofile.html). It's a small sampling profiler that is quick to setup and use on most systems. It's integrated into our build system for even greater convenience.
+
+Basic steps:
+1. Install gperftools. It's available in many package managers (apt packages `google-perftools` and `libgoogle-perftools-dev`, spack or homebrew package `gperftools`). E.g., on the DLR-SC hpda cluster (insert the compiler version you are using): `module load spack-user; spack install gperftools%gcc@13.1.0; module load gperftools`.
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/benchmarks/profiling.md:26]: I tested it on my workstation using apt packages and on the cluster using the description here. I think the missing step is loading `PrgEnv/gcc13-<something>`

[the remark this answers is not in the document]
Comment (reneSchm) [reply cpp/benchmarks/profiling.md:26]: Issue was spack specific.

File: cpp/benchmarks/profiling.md (hunk 1 above)
Comment (dabele) [reply cpp/benchmarks/profiling.md]: we are working on a full explanation. You are right that the wording was weird, improved it.

File: cpp/memilio/utils/profiler.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#ifndef MIO_UTILS_PROFILER_H
+#define MIO_UTILS_PROFILER_H
+
+#include "memilio/config.h"
+
+#ifdef MEMILIO_ENABLE_PROFILING
+
+#include "gperftools/profiler.h"
+#define MEMILIO_PROFILER_START(file) ProfilerStart(file)
```
[the remark this answers is not in the document]
Comment (dabele) [reply cpp/memilio/utils/profiler.h:36]: added documentation to the whole file.

**Comment by dabele** [comment]
added documentation for the cmake option (including some others that were added recently).

added links to profiling to the cpp README.

File: cpp/benchmarks/profiling.md [hunk 4]
```diff
@@ -23,8 +23,8 @@ How to use these tools, the general optimization loop:
...
 
 Basic steps:
-1. Install gperftools. It's available in many package managers (apt packages `google-perftools` and `libgoogle-perftools-dev`, spack or homebrew package `gperftools`). E.g., on the DLR-SC hpda cluster (insert the compiler version you are using): `module load spack-user; spack install gperftools%gcc@13.1.0; module load gperftools`.
-2. configure memilio with cmake variable `MEMILIO_ENABLE_PROFILING=ON`.
+1. Install gperftools. It's available in many package managers (apt packages `google-perftools` and `libgoogle-perftools-dev`, spack or homebrew package `gperftools`). E.g., on the DLR-SC hpda cluster (insert the compiler version you are using): `module load spack-user; module load PrgEnv/gcc13-openmpi; spack install gperftools%gcc@13.3.0; module load gperftools`.
+2. configure memilio with cmake variable `MEMILIO_ENABLE_PROFILING=ON`. Note that compiling with profiling enabled does not incur any runtime overhead unless profling is also enabled at runtime (see step 4), so developers can just enable it always for convenience.
```
Comment (reneSchm) [inline cpp/benchmarks/profiling.md:27]: > Note that compiling with profiling enabled does not incur any runtime overhead 

I wonder, is that also true for RelWithDebugInfo? Do they work the same?
(I am just curious, no changes needed)

Comment (dabele) [reply cpp/benchmarks/profiling.md:27]: I believe it is, the binary output of `RelWithDebInfo` should be the same as `Release` with some (maybe unused) source code positions.

**Review by reneSchm — APPROVED** [review]
Thanks for writing all this documentation!

I could verify the gperf steps work on native Linux, the Cluster *should* work as well, but had some issues possibly due to spack quirks.

**Review by DavidKerkmann — COMMENTED** [review]
I was able to run the profiler on MacOS 11.7.10.
It was important to also install the PkgConfig, I think we didn't have this as a requirement before. Perhaps we can add this to the installation process?
One thing that didn't work is the conversion of the .out file to .pdf, as proof-symbolize tries to access libs that were moved with MacOS 11. I am not sure how to resolve this. Here is some more information:

`error: /Library/Developer/CommandLineTools/usr/bin/otool-classic: can't open file: /usr/lib/liboah.dylib (No such file or directory)`
and about 100 similar more.

Here's a thread about this on apple.com:
https://developer.apple.com/forums/thread/655588?answerId=623718022#623718022

And another issue (thanks René) that lists the same problem:
https://github.com/google/pprof/issues/726

**Pull request merged**